pkg/sbom/cyclonedx/marshal.go

```diff
@@ -188,15 +188,64 @@ func externalRef(bomLink string, bomRef string) (string, error) {
 	return fmt.Sprintf("%s/%d#%s", strings.Replace(bomLink, "uuid", "cdx", 1), cdx.BOMFileFormatJSON, bomRef), nil
 }
 
+func generateDependencyGraph(sourceDependencies map[string][]string, idBomMap map[string]string) ([]cdx.Dependency, []cdx.Dependency) {
+	heads := []cdx.Dependency{}
+	deps := []cdx.Dependency{}
+
+	dependents := map[string]struct{}{}
+	bomDeps := map[string]*cdx.Dependency{}
+
+	for sourceId, sourceDependents := range sourceDependencies {
+		bomRef := idBomMap[sourceId]
+
+		dependencies := []cdx.Dependency{}
+		for _, sourceDepId := range sourceDependents {
+			dependentBomRef := idBomMap[sourceDepId]
+			if _, ok := bomDeps[dependentBomRef]; !ok {
+				bomDeps[dependentBomRef] = &cdx.Dependency{
+					Ref: dependentBomRef,
+				}
+			}
+			dependents[dependentBomRef] = struct{}{}
+			dependencies = append(dependencies, *bomDeps[dependentBomRef])
+		}
+		sort.SliceStable(dependencies, func(i, j int) bool {
+			return dependencies[i].Ref < dependencies[j].Ref
+		})
+
+		dependency := cdx.Dependency{
+			Ref:          bomRef,
+			Dependencies: &dependencies,
+		}
+		bomDeps[bomRef] = &dependency
+		deps = append(deps, dependency)
+	}
+
+	for _, depId := range maps.Keys(bomDeps) {
+		if _, ok := dependents[depId]; !ok {
+			heads = append(heads, *bomDeps[depId])
+		}
+	}
+	sort.SliceStable(heads, func(i, j int) bool {
+		return heads[i].Ref < heads[j].Ref
+	})
+	sort.SliceStable(deps, func(i, j int) bool {
+		return deps[i].Ref < deps[j].Ref
+	})
+
+	return heads, deps
+}
+
 func (e *Marshaler) marshalComponents(r types.Report, bomRef string) (*[]cdx.Component, *[]cdx.Dependency, *[]cdx.Vulnerability, error) {
 	var components []cdx.Component
-	var dependencies []cdx.Dependency
 	var metadataDependencies []cdx.Dependency
+	dependencies := map[string]cdx.Dependency{}
 	libraryUniqMap := map[string]struct{}{}
 	vulnMap := map[string]cdx.Vulnerability{}
 	for _, result := range r.Results {
 		bomRefMap := map[string]string{}
-		var componentDependencies []cdx.Dependency
+		bomPkgIDMap := map[string]string{}
```

Comment:
It looks unnecessarily complicated. Can we iterate packages and generate the mapping between package IDs and bom-refs first? The goal is simplifying `generateDependencyGraph`.

Reply:
ok, I'll try to do it